src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -2785,5 +2837,1166 @@ private void AssertValid()
+            }
+            else
+            {
+                // When the value is negative, we need to tzcnt the two's complement representation
+                // We'll do this "inline" to avoid needing to unnecessarily allocate.
+
+                uint part = ~value._bits[0] + 1;
+
+                for (int i = 1; (part == 0) && (i < value._bits.Length); i++)
+                {
+                    // Simply process bits, adding the carry while the previous value is zero
+
+                    part = ~value._bits[i] + 1;
+                    result += (sizeof(uint) * 8);
+
+                    i++;
+                }
+
+                result += uint.TrailingZeroCount(part);
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.GetShortestBitLength()" />
+        long IBinaryInteger<BigInteger>.GetShortestBitLength()
+        {
+            uint[]? bits = _bits;
+
+            if (bits is null)
+            {
+                int value = _sign;
+
+                if (value >= 0)
+                {
+                    return (sizeof(int) * 8) - int.LeadingZeroCount(value);
+                }
+                else
+                {
+                    return (sizeof(int) * 8) + 1 - int.LeadingZeroCount(~value);
+                }
+            }
+
+            long result = (bits.Length - 1) * 32;
+
+            if (_sign >= 0)
+            {
+                result += (sizeof(uint) * 8) - uint.LeadingZeroCount(bits[^1]);
+            }
+            else
+            {
+                result += (sizeof(uint) * 8) + 1 - uint.LeadingZeroCount(~bits[^1]);
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.GetByteCount()" />
+        int IBinaryInteger<BigInteger>.GetByteCount()
+        {
+            uint[]? bits = _bits;
+
+            if (bits is null)
+            {
+                return sizeof(int);
+            }
+            return bits.Length * 4;
+        }
+
+        /// <inheritdoc cref="IBinaryInteger{TSelf}.TryWriteLittleEndian(Span{byte}, out int)" />
+        bool IBinaryInteger<BigInteger>.TryWriteLittleEndian(Span<byte> destination, out int bytesWritten)
+        {
+            uint[]? bits = _bits;
+            int byteCount = (bits is null) ? sizeof(int) : bits.Length * 4;
+
+            if (destination.Length >= byteCount)
+            {
+                if (bits is null)
+                {
+                    int value = BitConverter.IsLittleEndian ? _sign : BinaryPrimitives.ReverseEndianness(_sign);
+                    Unsafe.WriteUnaligned(ref MemoryMarshal.GetReference(destination), value);
+                }
+                else if (_sign >= 0)
+                {
+                    // When the value is positive, we simply need to copy all bits as little endian
+
+                    ref byte address = ref MemoryMarshal.GetReference(destination);
```

Comment:
No need for `GetReference` in the common little-endian case. If the length check on line 3308 above succeeds, you know that `MemoryMarshal.Cast<uint, byte>(bits).CopyTo(destination)` will also succeed. The only time you'd need to bswap is in the big-endian case.

Once your two's-complement implementation comes online, you'll be able to reuse the same memcpy code there.

Reply:
I'm going to intentionally keep this simple for now and handle it as part of the rewrite.